File: bci/api/typescript/src/protocol/MessageBuilder.ts
/**
 * WIA BCI Message Builder
 *
 * Factory for creating protocol messages.
 */

import {
  BciMessage,
  MessageType,
  ConnectPayload,
  ConnectAckPayload,
  DisconnectPayload,
  StartStreamPayload,
  StopStreamPayload,
  StreamAckPayload,
  SignalPayload,
  SignalBatchPayload,
  MarkerPayload,
  CommandPayload,
  CommandAckPayload,
  ConfigPayload,
  StatusPayload,
  ErrorPayload,
  PingPayload,
  PongPayload,
  PROTOCOL_NAME,
  PROTOCOL_VERSION,
} from './types';

/**
 * Generate UUID v4
 */
function generateUuid(): string {
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
    const r = (Math.random() * 16) | 0;
    const v = c === 'x' ? r : (r & 0x3) | 0x8;
    return v.toString(16);
  });
}

/**
 * Message builder for WIA BCI protocol
 */
export class MessageBuilder {
  private sequence = 0;
  private sessionId?: string;

  /**
   * Set session ID for messages
   */
  setSessionId(sessionId: string): void {
    this.sessionId = sessionId;
  }

  /**
   * Reset sequence counter
   */
  resetSequence(): void {
    this.sequence = 0;
  }

  /**
   * Create base message
   */
  private createBase<T>(type: MessageType, payload: T): BciMessage<T> {
    const message: BciMessage<T> = {
      protocol: PROTOCOL_NAME,
      version: PROTOCOL_VERSION,
      messageId: generateUuid(),
      timestamp: Date.now(),
      type,
      payload,
    };

    if (this.sessionId) {
      message.sessionId = this.sessionId;
    }

    return message;
  }

  /**
   * Create base message with sequence
   */
  private createWithSequence<T>(type: MessageType, payload: T): BciMessage<T> {
    const message = this.createBase(type, payload);
    message.sequence = this.sequence++;
    return message;
  }

  // Connection messages

  /**
   * Create connect message
   */
  connect(payload: ConnectPayload): BciMessage<ConnectPayload> {
    return this.createBase('connect', payload);
  }

  /**
   * Create connect ack message
   */
  connectAck(payload: ConnectAckPayload): BciMessage<ConnectAckPayload> {
    if (payload.sessionId) {
      this.sessionId = payload.sessionId;
    }
    return this.createBase('connect_ack', payload);
  }

  /**
   * Create disconnect message
   */
  disconnect(payload: DisconnectPayload = {}): BciMessage<DisconnectPayload> {
    return this.createBase('disconnect', payload);
  }

  // Stream control messages

  /**
   * Create start stream message
   */
  startStream(payload: StartStreamPayload = {}): BciMessage<StartStreamPayload> {
    return this.createBase('start_stream', payload);
  }

  /**
   * Create stop stream message
   */
  stopStream(payload: StopStreamPayload = {}): BciMessage<StopStreamPayload> {
    return this.createBase('stop_stream', payload);
  }

  /**
   * Create stream ack message
   */
  streamAck(payload: StreamAckPayload): BciMessage<StreamAckPayload> {
    return this.createBase('stream_ack', payload);
  }

  // Signal messages

  /**
   * Create signal message
   */
  signal(payload: SignalPayload): BciMessage<SignalPayload> {
    return this.createWithSequence('signal', payload);
  }

  /**
   * Create signal batch message
   */
  signalBatch(payload: SignalBatchPayload): BciMessage<SignalBatchPayload> {
    return this.createWithSequence('signal_batch', payload);
  }

  // Marker message

  /**
   * Create marker message
   */
  marker(payload: MarkerPayload): BciMessage<MarkerPayload> {
    return this.createWithSequence('marker', payload);
  }

  // Command messages

  /**
   * Create command message
   */
  command(payload: CommandPayload): BciMessage<CommandPayload> {
    return this.createBase('command', payload);
  }

  /**
   * Create command ack message
   */
  commandAck(payload: CommandAckPayload): BciMessage<CommandAckPayload> {
    return this.createBase('command_ack', payload);
  }

  // Config message

  /**
   * Create config message
   */
  config(payload: ConfigPayload): BciMessage<ConfigPayload> {
    return this.createBase('config', payload);
  }

  // Status message

  /**
   * Create status message
   */
  status(payload: StatusPayload): BciMessage<StatusPayload> {
    return this.createBase('status', payload);
  }

  // Error message

  /**
   * Create error message
   */
  error(payload: ErrorPayload): BciMessage<ErrorPayload> {
    return this.createBase('error', payload);
  }

  // Heartbeat messages

  /**
   * Create ping message
   */
  ping(payload: PingPayload = {}): BciMessage<PingPayload> {
    return this.createBase('ping', {
      ...payload,
      clientTime: Date.now(),
    });
  }

  /**
   * Create pong message
   */
  pong(clientTime?: number): BciMessage<PongPayload> {
    return this.createBase('pong', {
      serverTime: Date.now(),
      clientTime,
    });
  }
}

/**
 * Default message builder instance
 */
export const messageBuilder = new MessageBuilder();

File: bci/api/python/wia_bci/protocol/message_builder.py
"""
WIA BCI Message Builder

Factory for creating protocol messages.
"""

import uuid
import time
from typing import Optional, Dict, Any, List

from .types import (
    BciMessage,
    MessageType,
    ConnectPayload,
    ConnectAckPayload,
    DisconnectPayload,
    StartStreamPayload,
    StopStreamPayload,
    StreamAckPayload,
    SignalPayload,
    SignalBatchPayload,
    MarkerPayload,
    CommandPayload,
    CommandAckPayload,
    ConfigPayload,
    StatusPayload,
    ErrorPayload,
    PingPayload,
    PongPayload,
    ConnectionState,
    PROTOCOL_NAME,
    PROTOCOL_VERSION,
)


def generate_uuid() -> str:
    """Generate UUID v4."""
    return str(uuid.uuid4())


def current_timestamp() -> int:
    """Get current timestamp in milliseconds."""
    return int(time.time() * 1000)


class MessageBuilder:
    """Message builder for WIA BCI protocol."""

    def __init__(self):
        self._sequence = 0
        self._session_id: Optional[str] = None

    def set_session_id(self, session_id: str) -> None:
        """Set session ID for messages."""
        self._session_id = session_id

    def reset_sequence(self) -> None:
        """Reset sequence counter."""
        self._sequence = 0

    def _create_base(self, msg_type: MessageType, payload: Dict[str, Any]) -> BciMessage:
        """Create base message."""
        return BciMessage(
            protocol=PROTOCOL_NAME,
            version=PROTOCOL_VERSION,
            message_id=generate_uuid(),
            timestamp=current_timestamp(),
            type=msg_type,
            payload=payload,
            session_id=self._session_id,
        )

    def _create_with_sequence(
        self, msg_type: MessageType, payload: Dict[str, Any]
    ) -> BciMessage:
        """Create message with sequence number."""
        message = self._create_base(msg_type, payload)
        message.sequence = self._sequence
        self._sequence += 1
        return message

    # Connection messages

    def connect(self, payload: ConnectPayload) -> BciMessage:
        """Create connect message."""
        return self._create_base(MessageType.CONNECT, payload.to_dict())

    def connect_ack(self, payload: ConnectAckPayload) -> BciMessage:
        """Create connect ack message."""
        if payload.session_id:
            self._session_id = payload.session_id
        return self._create_base(MessageType.CONNECT_ACK, payload.to_dict())

    def disconnect(
        self, payload: Optional[DisconnectPayload] = None
    ) -> BciMessage:
        """Create disconnect message."""
        p = payload or DisconnectPayload()
        return self._create_base(MessageType.DISCONNECT, p.to_dict())

    # Stream control messages

    def start_stream(
        self, payload: Optional[StartStreamPayload] = None
    ) -> BciMessage:
        """Create start stream message."""
        p = payload or StartStreamPayload()
        return self._create_base(MessageType.START_STREAM, p.to_dict())

    def stop_stream(
        self, payload: Optional[StopStreamPayload] = None
    ) -> BciMessage:
        """Create stop stream message."""
        p = payload or StopStreamPayload()
        return self._create_base(MessageType.STOP_STREAM, p.to_dict())

    def stream_ack(self, payload: StreamAckPayload) -> BciMessage:
        """Create stream ack message."""
        return self._create_base(MessageType.STREAM_ACK, payload.to_dict())

    # Signal messages

    def signal(self, payload: SignalPayload) -> BciMessage:
        """Create signal message."""
        return self._create_with_sequence(MessageType.SIGNAL, payload.to_dict())

    def signal_batch(self, payload: SignalBatchPayload) -> BciMessage:
        """Create signal batch message."""
        return self._create_with_sequence(MessageType.SIGNAL_BATCH, payload.to_dict())

    # Marker message

    def marker(self, payload: MarkerPayload) -> BciMessage:
        """Create marker message."""
        return self._create_with_sequence(MessageType.MARKER, payload.to_dict())

    # Command messages

    def command(self, payload: CommandPayload) -> BciMessage:
        """Create command message."""
        return self._create_base(MessageType.COMMAND, payload.to_dict())

    def command_ack(self, payload: CommandAckPayload) -> BciMessage:
        """Create command ack message."""
        return self._create_base(MessageType.COMMAND_ACK, payload.to_dict())

    # Config message

    def config(self, payload: ConfigPayload) -> BciMessage:
        """Create config message."""
        return self._create_base(MessageType.CONFIG, payload.to_dict())

    # Status message

    def status(self, payload: StatusPayload) -> BciMessage:
        """Create status message."""
        return self._create_base(MessageType.STATUS, payload.to_dict())

    # Error message

    def error(self, payload: ErrorPayload) -> BciMessage:
        """Create error message."""
        return self._create_base(MessageType.ERROR, payload.to_dict())

    # Heartbeat messages

    def ping(self, payload: Optional[PingPayload] = None) -> BciMessage:
        """Create ping message."""
        p = payload or PingPayload(client_time=current_timestamp())
        return self._create_base(MessageType.PING, p.to_dict())

    def pong(self, client_time: Optional[int] = None) -> BciMessage:
        """Create pong message."""
        payload = PongPayload(
            server_time=current_timestamp(), client_time=client_time
        )
        return self._create_base(MessageType.PONG, payload.to_dict())


# Default message builder instance
message_builder = MessageBuilder()
